### PR TITLE
Basic implementation of PKCE

### PR DESCRIPTION
**Pull request by Teeed**
This is my own implementation of PKCE, it is similar to #1407 but this version could be directly merged.

Issue #1114

### CODE REVIEW

**Comment by mfmarche** [comment]
Hi @Teeed , I tried your PR and it was mostly functional.

It seems there is no way to get to the handleAuthCode without actually passing the basic auth (providing a valid client secret). Shouldn't the handling of the secret be deferred only in the case that PKCE flow is not used (ie. when there is no code_verifier?). Making this change allows for PKCE to commence without a secret.

**Comment by justin-slowik** [comment]
Has there been any progress on implementing PKCE?  I've been working on a branch which implements the OAuth 2 Device Authorization Grant, and I know that to make a fully featured native app flow, PKCE should be included as well.  Would be beneficial not to duplicate work

File: server/handlers.go [hunk 1]
```diff
@@ -777,6 +798,22 @@ func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client s
...
+	if codeChallengeFromStorage != "" {
+		providedCodeVerifier := r.PostFormValue("code_verifier")
+		calculatedCodeChallenge, err := s.calculateCodeChallenge(providedCodeVerifier, authCode.CodeChallenge.CodeChallengeMethod)
+		if err != nil {
+			s.tokenErrHelper(w, errServerError, "", http.StatusInternalServerError)
+			return
+		}
+
+		if codeChallengeFromStorage != calculatedCodeChallenge {
+			s.tokenErrHelper(w, errInvalidRequest, "invalid code_verifier.", http.StatusBadRequest)
```
Comment (deric) [inline server/handlers.go:812]: According to [RFC 7636](https://tools.ietf.org/html/rfc7636#section-4.6), `errInvalidGrant` should be returned.

 > If the values are not equal, an error response indicating "invalid_grant" as described in Section 5.2 of [RFC6749] MUST be returned.

**Comment by Teeed** [comment]
Yeah. First of all: sorry for not responding for such a long time.

Company which I work for (and which allowed me to push these changes to upstream) decided that this changes are no longer needed and hence development of these changes was stalled. I currently work on different projects, so I do not have time (and business reason :() to fix raised issues.

If anyone wants to continue this work - please do!

**Comment by heidemn-faro** [comment]
Hi @Teeed, thanks for your efforts here.
Please allow me a stupid question, since I'm not familiar with the codebase:

Does this add:
1. PKCE for the app (e.g. a single-page app) connecting to Dex, or
2. PKCE for Dex connecting to the upstream IdP?

We would be interested in case 1.

**Comment by heidemn-faro** [comment]
Hi @Teeed , 

> If anyone wants to continue this work - please do!

Our company wants to finish your work.
Could you please sign off your commit, meaning that you accept the Dex [DCO](https://github.com/dexidp/dex/blob/master/DCO)?
Otherwise, I'm not sure if we're allowed to reuse your work in our follow-up PR.

Signing off the commit should work like this:

```
git checkout <BRANCH>
git commit --amend --signoff
git push --force-with-lease <ORIGIN> <BRANCH>
```

Thanks

**Comment by HEllRZA** [comment]
This PR can be closed as #1784, which is based on this, has been merged. Thanks again to @Teeed for his work!

**Comment by Teeed** [comment]
Closing, as these changes got merged with #1784

**Pull request closed without merging**